File: backend/Models/CitizenRequest.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;
using NetTopologySuite.Geometries;

namespace MvdBackend.Models
{
    public class CitizenRequest
    {
        public int Id { get; set; }

        [ForeignKey(nameof(Citizen))]
        public int CitizenId { get; set; }

        [JsonIgnore]
        public Citizen Citizen { get; set; } = null!;

        [ForeignKey(nameof(RequestType))]
        public int RequestTypeId { get; set; }

        [JsonIgnore]
        public RequestType RequestType { get; set; } = null!;

        [ForeignKey(nameof(Category))]
        public int CategoryId { get; set; }

        [JsonIgnore]
        public Category Category { get; set; } = null!;

        [Required(ErrorMessage = "Описание обязательно")]
        [StringLength(5000, MinimumLength = 10, ErrorMessage = "Описание должно быть от 10 до 5000 символов")]
        public string Description { get; set; } = "";

        [ForeignKey(nameof(AcceptedBy))]
        public int AcceptedById { get; set; }

        [JsonIgnore]
        public Employee AcceptedBy { get; set; } = null!;

        [ForeignKey(nameof(AssignedTo))]
        public int? AssignedToId { get; set; }

        [JsonIgnore]
        public Employee? AssignedTo { get; set; }

        [Required]
        public DateTime IncidentTime { get; set; }
        
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        [Required(ErrorMessage = "Адрес инцидента обязателен")]
        [StringLength(500, ErrorMessage = "Адрес не должен превышать 500 символов")]
        public string IncidentLocation { get; set; } = "";

        [StringLength(500, ErrorMessage = "Адрес не должен превышать 500 символов")]
        public string CitizenLocation { get; set; } = "";

        [Required]
        [StringLength(20)]
        public string RequestNumber { get; set; } = "";
        
        public DateTime? UpdatedAt { get; set; }

        [ForeignKey(nameof(RequestStatus))]
        public int RequestStatusId { get; set; }

        [JsonIgnore]
        public RequestStatus RequestStatus { get; set; } = null!;

        [ForeignKey(nameof(District))]
        public int? DistrictId { get; set; }

        [JsonIgnore]
        public District? District { get; set; }

        // PostGIS Point для географических координат
        public Point? Location { get; set; }

        // AI-анализ
        [StringLength(200)]
        public string? AiCategory { get; set; }

        [StringLength(50)]
        public string? AiPriority { get; set; }

        [StringLength(1000)]
        public string? AiSummary { get; set; }

        [StringLength(1000)]
        public string? AiSuggestedAction { get; set; }

        [StringLength(50)]
        public string? AiSentiment { get; set; }

        public DateTime? AiAnalyzedAt { get; set; }

        public bool IsAiCorrected { get; set; }

        [StringLength(200)]
        public string? FinalCategory { get; set; }
    }
}

File: frontend-user/lib/models/citizen_request.dart
class CitizenRequestDto {
  final int id;
  final int citizenId;
  final int requestTypeId;
  final int categoryId;
  final String description;
  final int acceptedById;
  final int? assignedToId;
  final String incidentTime;
  final String createdAt;
  final String incidentLocation;
  final String citizenLocation;
  final int requestStatusId;
  final int? districtId;
  final String requestNumber;
  final double? latitude;
  final double? longitude;
  final String? aiCategory;
  final String? aiPriority;
  final String? aiSummary;
  final String? aiSuggestedAction;
  final String? aiSentiment;
  final String? aiAnalyzedAt;
  final bool? isAiCorrected;
  final String? finalCategory;

  const CitizenRequestDto({
    required this.id,
    required this.citizenId,
    required this.requestTypeId,
    required this.categoryId,
    required this.description,
    required this.acceptedById,
    required this.assignedToId,
    required this.incidentTime,
    required this.createdAt,
    required this.incidentLocation,
    required this.citizenLocation,
    required this.requestStatusId,
    required this.districtId,
    required this.requestNumber,
    required this.latitude,
    required this.longitude,
    required this.aiCategory,
    required this.aiPriority,
    required this.aiSummary,
    required this.aiSuggestedAction,
    required this.aiSentiment,
    required this.aiAnalyzedAt,
    required this.isAiCorrected,
    required this.finalCategory,
  });

  factory CitizenRequestDto.fromJson(Map<String, dynamic> json) {
    try {
      return CitizenRequestDto(
        id: json['id'] as int,
        citizenId: json['citizenId'] as int,
        requestTypeId: json['requestTypeId'] as int,
        categoryId: json['categoryId'] as int,
        description: json['description'] as String? ?? '',
        acceptedById: json['acceptedById'] as int,
        assignedToId: json['assignedToId'] as int?,
        incidentTime: json['incidentTime'] as String? ?? DateTime.now().toIso8601String(),
        createdAt: json['createdAt'] as String? ?? DateTime.now().toIso8601String(),
        incidentLocation: json['incidentLocation'] as String? ?? '',
        citizenLocation: json['citizenLocation'] as String? ?? '',
        requestStatusId: json['requestStatusId'] as int? ?? 1,
        districtId: json['districtId'] as int?,
        requestNumber: json['requestNumber'] as String? ?? '',
        latitude: (json['latitude'] as num?)?.toDouble(),
        longitude: (json['longitude'] as num?)?.toDouble(),
        aiCategory: json['aiCategory'] as String?,
        aiPriority: json['aiPriority'] as String?,
        aiSummary: json['aiSummary'] as String?,
        aiSuggestedAction: json['aiSuggestedAction'] as String?,
        aiSentiment: json['aiSentiment'] as String?,
        aiAnalyzedAt: json['aiAnalyzedAt'] as String?,
        isAiCorrected: json['isAiCorrected'] as bool? ?? false,
        finalCategory: json['finalCategory'] as String?,
      );
    } catch (e) {
      throw FormatException('Ошибка парсинга CitizenRequestDto: $e');
    }
  }
}

class CreateCitizenRequestDto {
  final int citizenId;
  final int requestTypeId;
  final int categoryId;
  final String description;
  final int acceptedById;
  final int? assignedToId;
  final String incidentTime; // ISO 8601
  final String incidentLocation;
  final String citizenLocation;
  final double? latitude;
  final double? longitude;
  final int requestStatusId;

  const CreateCitizenRequestDto({
    required this.citizenId,
    required this.requestTypeId,
    required this.categoryId,
    required this.description,
    required this.acceptedById,
    this.assignedToId,
    required this.incidentTime,
    required this.incidentLocation,
    required this.citizenLocation,
    this.latitude,
    this.longitude,
    required this.requestStatusId,
  });

  Map<String, dynamic> toJson() => {
        'citizenId': citizenId,
        'requestTypeId': requestTypeId,
        'categoryId': categoryId,
        'description': description,
        'acceptedById': acceptedById,
        if (assignedToId != null) 'assignedToId': assignedToId,
        'incidentTime': incidentTime,
        'incidentLocation': incidentLocation,
        'citizenLocation': citizenLocation,
        if (latitude != null) 'latitude': latitude,
        if (longitude != null) 'longitude': longitude,
        'requestStatusId': requestStatusId,
      };

  }






